### PR TITLE
bots: Adapt push-rewrite to webhook quickly posting test statuses

### PR DESCRIPTION
**Pull request by Gundersanne**
TODO:
- [x] test this on this very pr!

### CODE REVIEW

File: bots/push-rewrite [hunk 1]
```diff
@@ -54,6 +61,10 @@ def main():
     api = github.GitHub(repo=opts.repo)
     old_statuses = api.statuses(remote)
 
+    # add [no-test] to title to prevent tests bring triggered
+    pull = find_pr_with_sha(api, remote)
+    api.post("pulls/{0}".format(pull["number"]), "{0} [no-test]".format(pull["title"]))
```
Comment (martinpitt) [inline bots/push-rewrite]: As this works on an existing PR by definition, couldn't this now use the label? This would avoid github notification email noise.

Comment (Gundersanne) [reply bots/push-rewrite]: oh yea, then that's preferred

File: bots/push-rewrite [hunk 2]
```diff
@@ -30,6 +30,13 @@ def execute(*args):
 def git(*args):
     return execute("git", *args).strip()
 
+def find_pr_with_sha(api, sha):
+    pulls = api.pulls()
+    for pull in pulls:
+        if pull["head"]["sha"] == sha:
+            return pull
```
Comment (martinpitt) [inline bots/push-rewrite]: Too bad that the [commits API](https://api.github.com/repos/cockpit-project/cockpit/commits/a935533b4afae934dad4d8cb13a9ecc4bfdf49c9) doesn't refer back to the PR.. So I suppose this is necessary. But this should have some error handling if the PR isn't found.

File: bots/push-rewrite [hunk 3]
```diff
@@ -54,6 +68,13 @@ def main():
     api = github.GitHub(repo=opts.repo)
     old_statuses = api.statuses(remote)
 
+    # add [no-test] to title to prevent tests bring triggered
+    pull = find_pr_with_sha(api, remote)
+    labels = labels_of_pr(pull)
+    tests_enabled = "no-test" in labels or "[no-test]" in pull["title"]
+    if tests_enabled:
+        api.post("issues/{0}/labels".format(pull["number"]), labels  + ["no-test"])
```
Comment (martinpitt) [inline bots/push-rewrite]: This you can do with `label()` from `bots/task/__init__.py`.

File: bots/push-rewrite [hunk 4]
```diff
@@ -30,6 +30,20 @@ def execute(*args):
...
+    for pull in pulls:
+        if pull["head"]["sha"] == sha:
+            return pull
+    sys.stderr.write("Could not find pull with revision {0}.\n".format(sha))
+    return 1
+
+def labels_of_pr(pull):
+    if "labels" not in pull:
+        pull["labels"] = api.get("issues/{0}/labels".format(number))
+    return list(map(lambda label: label["name"], pull["labels"]))
```
Comment (martinpitt) [inline bots/push-rewrite]: This duplicates code  from bots/tests-scan. Can this please move into bots/task/github.py?

**Comment by Gundersanne** [comment]
So if the external label is added it does add those, but that's to be expected. Maybe we should change that too.

File: bots/push-rewrite [hunk 5]
```diff
@@ -54,6 +62,13 @@ def main():
     api = github.GitHub(repo=opts.repo)
     old_statuses = api.statuses(remote)
 
+    pull = find_pr_with_sha(api, remote)
+    labels = labels_of_pull(pull)
+    tests_enabled = "no-test" in labels or "[no-test]" in pull["title"]
+    # add no-test label to prevent tests bring triggered
+    if tests_enabled:
+        label(pull, labels + ["no-test"])
```
Comment (martinpitt) [inline bots/push-rewrite]: This should be `tests_disabled`. However, why is that necessary? If the test already  has "no-test", then there's no need to do it again.

It seems to me that the condition for `tests_enabled` should be negated, so that this labels/unlabels the PR if `no-test` is *not* present?

Comment (Gundersanne) [reply bots/push-rewrite]: oh you're right, must've tested it then without removing the label. Fixing this.

File: bots/push-rewrite [hunk 6]
```diff
@@ -30,6 +30,14 @@ def execute(*args):
 def git(*args):
     return execute("git", *args).strip()
 
+def find_pr_with_sha(api, sha):
+    pulls = api.pulls()
+    for pull in pulls:
+        if pull["head"]["sha"] == sha:
+            return pull
+    sys.stderr.write("Could not find pull with revision {0}.\n".format(sha))
+    return 1
```
Comment (martinpitt) [inline bots/push-rewrite]: This will just be interpreted by the  caller as "pull request #1", so please rather do sys.exit() here.
